[CICD] Add CLI Command Test Coverage

wangzelin007 opened the pull request:
Description:

- `azdev cmdcov` atuo generate CLI Command Test Coverage Report about every module to track and improve command-level test coverage.
- `azdev linter` add rules to detect any missing test cases in every pull request and continuously improve code quality. 

Examples

    Check all CLI modules command test coverage.
        azdev cmdcov CLI

    Check one or serveral modules command test coverage.
        azdev cmdcov vm storage

    Check CLI modules command argument test coverage.
        azdev cmdcov CLI --level argument

    Add linter rules to check every PR
        azdev linter --rule-types command_test_coverage --repo=repo_path --src=src_branch --tgt=target_branch
Report:

![image](https://user-images.githubusercontent.com/18628534/145713636-fadb6702-a1e8-4a6b-b0a1-407cccb27574.png)
![image](https://user-images.githubusercontent.com/18628534/145713733-25fcbb8a-5650-41b0-b34f-09d914971e25.png)
![image](https://user-images.githubusercontent.com/18628534/146549936-8c54dab1-8286-436f-8da9-a62acd4af9d5.png)

wangzelin007 committed:
add command coverage

wangzelin007 committed:
commit

wangzelin007 committed:
regex

wangzelin007 committed:
command coverage 1.0

wangzelin007 committed:
upload html

wangzelin007 committed:
modify report

wangzelin007 committed:
refactor

wangzelin007 committed:
add annotation

wangzelin007 committed:
Update test_cmdcov.py

wangzelin007 committed:
modify detect command function

wangzelin007 committed:
detect from live test file

wangzelin007 committed:
update func _get_all_commands

wangzelin007 committed:
Update __init__.py

wangzelin007 committed:
update sort function

wangzelin007 committed:
update sort function

wangzelin007 committed:
add level argument

wangzelin007 committed:
add cli own

wangzelin007 committed:
add EXCLUDE MODULES

wangzelin007 committed:
Update component.css

wangzelin007 committed:
delete unused code

wangzelin007 committed:
delete unused file

wangzelin007 committed:
modify regex

wangzelin007 committed:
add help info

wangzelin007 committed:
add description

wangzelin007 committed:
exclude command wait

wangzelin007 committed:
detect from recording file

wangzelin007 committed:
delete exception already fixed

wangzelin007 committed:
add newline

wangzelin007 committed:
add license headers

wangzelin007 committed:
fix tox

wangzelin007 committed:
add linter rule comcov

wangzelin007 committed:
refactoring

wangzelin007 committed:
update

wangzelin007 committed:
fix bug

wangzelin007 committed:
regex

wangzelin007 committed:
add log add exec_state

wangzelin007 committed:
exclude deprecated commands

wangzelin007 committed:
fix tox

wangzelin007 committed:
fix tox

wangzelin007 committed:
add rule type command_coverage

wangzelin007 committed:
add network exclude commands

wangzelin007 committed:
add number sign exclude

wangzelin007 committed:
update diff branches

wangzelin007 committed:
support exclusion

StrawnSC commented:
Hey is this currently functional? When I run any `azdev cmdcov` command (such as `azdev cmdcov -h`), I get the following exception: 
```ModuleNotFoundError: No module named 'azdev.operations.cmdcov'```
What command are you using to build `azdev`? I'm just doing a `pip install .` in the azdev repo, which worked for me on the dev branch

wangzelin007 commented:
> Hey is this currently functional? When I run any `azdev cmdcov` command (such as `azdev cmdcov -h`), I get the following exception: `ModuleNotFoundError: No module named 'azdev.operations.cmdcov'` What command are you using to build `azdev`? I'm just doing a `pip install .` in the azdev repo, which worked for me on the dev branch

@StrawnSC,
This feature currently is not functional, we will release a new version for it.
You can try it by replace the azdev code with this [branch](https://github.com/wangzelin007/azure-cli-dev-tools/tree/cmdcov).

wangzelin007 committed:
linter refactoring

wangzelin007 committed:
add comment

wangzelin007 committed:
move constant and regex out

wangzelin007 committed:
suppport linter_exclusion.yml in cmdcov

wangzelin007 committed:
add license headers

wangzelin007 committed:
fix pylint

wangzelin007 committed:
fix flake8

wangzelin007 committed:
Update test_cmdcov.py

wangzelin007 committed:
fix bug for linter rule

wangzelin007 committed:
add resource skip commands

wangzelin007 committed:
update

1. Change the selected color to pink
2. Mouse hover shows percentage details
3. Modify Coverage Report description

wangzelin007 committed:
role commands exclusion

wangzelin007 committed:
modify HISTORY.rst and setup.py

wangzelin007 committed:
resource skip two commands - hard to test

wangzelin007 committed:
bug fixes: search argument

wangzelin007 committed:
modify command coverage to command test coverage

wangzelin007 committed:
bug fixes: Add three new scenarios

wangzelin007 committed:
update CLI own and exclude commands

wangzelin007 committed:
move private dns out of cli own

wangzelin007 committed:
add log for exclusions

jsntcy commented:
@wangzelin007, Do we support adding `--level` for specific modules such as `azdev cmdcov storage --level argument`?

wangzelin007 committed:
exclude vm host restart

jsntcy commented on azdev/operations/constant.py:
```diff
@@ -0,0 +1,158 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# -----------------------------------------------------------------------------
```
Better to put this file (data) in azure-cli repo; otherwise we need release a new version for azdev if any update in this file. Decouple data from logic.

wangzelin007 replied:
May I ask where should I put this file ？
Can I put this file in here? 
`azure-cli\src\azure-cli\cmdcov_config.py`

wangzelin007 commented:
> @wangzelin007, Do we support adding `--level` for specific modules such as `azdev cmdcov storage --level argument`?

Yes~

jsntcy commented:
General guideline: Avoid to release a new version for the package when there is any changes.

wangzelin007 commented:
> General guideline: Avoid to release a new version for the package when there is any changes.

Sure, I will double check.
Can you recommend a path in azcli repo ?

wangzelin007 committed:
config

wangzelin007 commented:
/azp run

wangzelin007 committed:
Update test_config.py

wangzelin007 commented:
> Avoid to release a new version for the package when there is any changes.

Move CLI command test coverage config to azure-cli repo：https://github.com/Azure/azure-cli/pull/25441

wangzelin007 commented:
/azp run

wangzelin007 committed:
update

wangzelin007 committed:
update

wangzelin007 committed:
update
